### DISCUSSION TITLE
rOpenScPCA package: Ideas for functionality

### DISCUSSION THREAD

**Discussion by sjspielman** [AlexsLemonade/OpenScPCA-analysis#731] [Propose a new analysis]
### Proposed analysis

We have discussed (https://github.com/AlexsLemonade/OpenScPCA-analysis/discussions/727#discussioncomment-10488085) creating an R package to house functions for general use across OpenScPCA modules. ~It will be called `rOpenScPCA`, and it will live in the top-level of the repo in a `packages` directory. (note that this naming scheme keeps the door open to one day perhaps have a python package too and name it `pyOpenScPCA`!)~ 
(edit) `rOpenScPCA` now lives in its own repo [`AlexsLemonade/rOpenScPCA`](https://github.com/AlexsLemonade/rOpenScPCA). However, discussion for this package will continue _here_ in this repo!

To begin, this package will contain clustering utility functions as discussed in #727 (edited, wrong link originally).

Please use this discussion to generate ideas for functions that might go into the `rOpenScPCA` package! 

### Scientific goals

Design an R package with common functions for use throughout OpenScPCA R code.

### Methods or approach

R package.

### Existing modules

It will be used by other modules.

**Comment by cansavvy** [comment]
I'd love to be involved with this.

**Reply by sjspielman** [reply]
Yay! Right now I'm working on a new module `hello-clusters` to demonstrate usage of the existing functions in here. We don't have any specific plans yet for more functionality, and I imagine we won't get back to thinking about it until #786 is all wrapped up. In the meantime, let us know any thoughts you have for functions for the package! 😄

**Reply by sjspielman** [reply]
We've discussed a couple ideas for functions we might like to implement in `rOpenScPCA`, based on common tasks project contributors have performed that might be nice to offer a consistent approach for. Anything you might be interested in?

*Also fyi!!* The package moved and now lives here: https://github.com/AlexsLemonade/rOpenScPCA. For context, see #854.

- Convert gene symbols in an object
	- OpenScPCA objects use ensembl ids as feature identifiers, but some tasks with Seurat require gene symbols instead. A function to consistently support this conversion between ids would be useful. 
	- We could handle duplicates in two possible ways:
		- First, we should implement use our preferred approach: sum duplicates to get the total transcriptional output for a gene symbol
		- Second, we might also [implement Seurat's approach](https://github.com/satijalab/azimuth/blob/master/R/helpers.R) (I've linked Azimuth code here), which we think takes the first matching id symbol from a list, but we are admittedly thin on the specifics here.
	- We might begin with a function to just convert symbols/ids, and then a function to update rownames in a given object (Seurat or SCE).
- Add Seurat object conversion
	- A lot of folks prefer to work with Seurat pipelines, so we could offer support for them to convert SCEs to Seurat objects
	- An option question is how much do we "do" in this conversion? Should we just port over the bare minimum to create a Seurat object and then let users handle all Seurat processing on their own (let's call this "plain conversion"). Or, do we do something more involved and run down some Seurat pipeline steps from raw counts in the new object? Open for discussion how far we take this, but at a minimum a plain conversion is good to have an option for.
	- We did discuss this function could have an option to convert from ensembl -> gene symbols too, or just port in reduced dimension stuff which can be useful for downstream.

**Reply by sjspielman** [reply]
Woops, meant to also tag you ☝️ ! @cansavvy

**Reply by sjspielman** [reply]
Looks like ~simultaneous to my comment above, @jashapiro was opening this issue! https://github.com/AlexsLemonade/rOpenScPCA/issues/2

**Reply by cansavvy** [reply]
I haven't really used GitHub discussions much yet so I definitely didn't see your replies until now. Whoops!

**Comment by jashapiro** [comment]
I just wrote up https://github.com/AlexsLemonade/rOpenScPCA/issues/2 in our new repository for this package, describing how I envision a function (or small set of functions) to perform gene id conversion within our project. I see now that both @sjspielman and I were writing up our thoughts at the same time!
